state/state.go

```diff
@@ -1461,52 +1462,22 @@ func (st *State) processCommonModelApplicationArgs(args *AddApplicationArgs) (Ba
 	// handled by the client, so args.Release is not necessarily
 	// one of the charm's supported series. We require that the
 	// specified series is of the same operating system as one of
-	// the supported series. For old-style charms with the series
-	// in the URL, that series is the one and only supported
-	// series.
+	// the supported series.
 	appBase, err := corebase.ParseBase(args.CharmOrigin.Platform.OS, args.CharmOrigin.Platform.Channel)
```

Comment:
This block of code is again duplicating concepts we check in other places. Let's consolidate the logic, rather than have separate implementations all over.

Reply:
I have created `BaseIsCompatibleWithCharm` and `OSIsCompatibleWithCharm` in core/charm to consolidate this logic

There are many mroe places this can be rolled out, but for the time being I'm leaving just implemented in a few places this PR touches to avoid scope creep

Reply:
This raises an interesting question. Why are we checking the operating system type only at the state level? Did we remove the other checks and forget state? 

There was a conversation recently about using a centos charm on a different version of linux - where it should work. 

Please keep this in mind while working on the rest of this project. We should address in the near future.